nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,492 @@
+      --exclude-from ${mkExcludeFile cfg} \
+      ${cfg.extraCreateArgs} \
+      "::$archiveName" \
+      ${mkPathsArgs cfg}
+    ${cfg.postCreate}
+  '' + optionalString (cfg.prune.keep != { }) ''
+    borg prune \
+      ${mkKeepArgs cfg} \
+      ${cfg.extraPruneArgs}
+    ${cfg.postPrune}
+  '' + ''
+    ${cfg.postHook}
+  '';
+
+  mkPassEnv = cfg: with cfg.encryption;
+    if passCommand != null then
+      { BORG_PASSCOMMAND = passCommand; }
+    else if passphrase != null then
+      { BORG_PASSPHRASE = passphrase; }
+    else { };
+
+  mkBackupService = name: cfg: 
+    let
+      userHome = config.users.users.${cfg.user}.home;
+    in nameValuePair "borgbackup-job-${name}" {
+      description = "BorgBackup job ${name}";
+      path = with pkgs; [
+        borgbackup openssh
+      ];
+      script = mkBackupScript cfg;
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        # Only run when no other process is using CPU or disk
+        CPUSchedulingPolicy = "idle";
+        IOSchedulingClass = "idle";
+        ProtectSystem = "full";
+        #FIXME:
+        #ProtectSystem = "strict";
+        #ReadWritePaths =
+        #  [ "${userHome}/.config/borg" "${userHome}/.cache/borg" ]
+        #  # Borg needs write access to repo if it is not remote
+        #  ++ optional (isPath cfg.repo) cfg.repo;
+      };
+      environment = {
+        BORG_REPO = cfg.repo;
+        inherit (cfg) archiveName;
+      } // (mkPassEnv cfg) // cfg.environment;
+      inherit (cfg) startAt;
+    };
+
+  # The name arg is needed for mapAttrsToList
+  mkPassAssertion = name: cfg: {
+    assertion = with cfg.encryption;
+      mode != "none" -> passCommand != null || passphrase != null;
+    message =
+      ''passCommand or passphrase have to be specified ''
+      + ''because borgbackup.jobs.${name}.encryption != "none"'';
+  };
+
+  mkRepoService = name: cfg:
+    nameValuePair "borgbackup-repo-${name}" {
+      description = "Create BorgBackup repository ${name} directory";
+      script = ''
+        mkdir -p ${escapeShellArg cfg.path}
+        chown ${cfg.user}:${cfg.group} ${escapeShellArg cfg.path}
+      '';
+      serviceConfig = {
+        # The service's only task is to ensure that the specified path exists
+        Type = "oneshot";
+      };
+      wantedBy = [ "multi-user.target" ];
+    };
+
+  mkAuthorizedKey = cfg: appendOnly: key:
+    let
+      restrictedArg = "--restrict-to-repository ${escapeShellArg cfg.path}";
+      appendOnlyArg = optionalString appendOnly "--append-only";
+      quotaArg = optionalString (cfg.quota != null) "--storage-quota ${cfg.quota}";
+      command = "borg serve ${restrictedArg} ${appendOnlyArg} ${quotaArg}";
+    in
+      ''command="${command}",restrict ${key}'';
+
+  mkUsersConfig = name: cfg: {
+    users.${cfg.user} = {
+      openssh.authorizedKeys.keyFiles = singleton
+        (pkgs.writeText "borgbackup.${cfg.user}.authorized_keys" (concatStringsSep "\n"
```

Comment:
You can pass strings directly to "openssh.authorizedKeys.keys" instead of writing files and passing them to "...keyFiles".

Reply:
I used `keyFiles` because I wasn't able to merge multiple `keys` options using `mkMerge`. But I tried again and it worked :D